crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs

```diff
@@ -123,6 +118,81 @@ impl ArrayHeap {
         self.staged.insert(index, (value, op));
     }
 
+    /// This helper function transforms an expression into a linear expression, by generating a witness if the input expression is not linear nor constant
+    fn normalize_expression(expr: &Expression, evaluator: &mut Evaluator) -> Expression {
+        if !expr.is_linear() || expr.linear_combinations.len() > 1 {
+            let w = evaluator.create_intermediate_variable(expr.clone());
+            Expression::from(w)
+        } else {
+            expr.clone()
+        }
+    }
+
+    /// Decide which opcode to use, depending on the backend support
+    fn acir_gen(
+        &self,
+        evaluator: &mut Evaluator,
+        array_id: ArrayId,
+        array_len: u32,
+        is_opcode_supported: IsOpcodeSupported,
+    ) {
+        let dummy = MemoryBlock { id: AcirBlockId(0), len: 0, trace: Vec::new() };
+        let dummy2 = dummy.clone();
+        let dummy3 = dummy.clone();
+        let (block, ram, rom) = (
+            is_opcode_supported(&AcirOpcode::Block(dummy)),
+            is_opcode_supported(&AcirOpcode::RAM(dummy2)),
+            is_opcode_supported(&AcirOpcode::ROM(dummy3)),
+        );
+        if rom {
+            // If the backend support ROM and the array is read-only, we generate the ROM opcode
+            if let ArrayType::ReadOnly(len) = self.typ {
+                self.add_rom_opcode(evaluator, array_id, len.unwrap() as u32);
+                return;
+            }
+        }
+
+        match (block, ram) {
+            (false, false) => self.generate_permutation_constraints(evaluator, array_id, array_len),
+            (false, true) => self.add_ram_opcode(evaluator, array_id, array_len),
+            (true, _) => {
+                evaluator.opcodes.push(AcirOpcode::Block(MemoryBlock {
+                    id: AcirBlockId(array_id.as_u32()),
+                    len: array_len,
+                    trace: self.trace.clone(),
+                }));
+            }
+        }
```

Comment:
I think this would be cleaner if we just use `if-elses` here rather than having the `match`. You're essentially reconstructing the `if-else` logic with all the boolean flags here and an explicit `if-else` shows the "fallback" nature more clearly.

We could then inline the `is_opcode_supported` calls into the if conditions.